target/linux/ramips/dts/mt7620n_hootoo_tm05.dts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Device Tree file for the HooToo TM-05
+ *
+ * Copyright (C) 2020 Russell Morris (rmorris@rkmorris.us)
+ * 
+ * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+/dts-v1/;
+
+#include "mt7620n.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "hootoo,tm05", "ralink,mt7620n-soc";
+	model = "HooToo TM05";
+
+	aliases {
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+		label-mac-device = &ethernet;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_power: power {
+			label = "tm05:blue:power";
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		wifi {
+			label = "tm05:green:wifi";
+			gpios = <&gpio2 0 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "phy0tpt";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio2 1 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+};
+
+&gpio2 {
+	status = "okay";
+};
+
+&gpio3 {
+	status = "okay";
+};
+
+&i2c {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				label = "loader";
+				reg = <0x50000 0x1000>;
+				read-only;
+			};
+
+			partition@51000 {
```

Comment:
Are you able to sysupgrade with this offset? AFAIK firmware partition should start on block boundary.

Reply:
I did run a sysupgrade, it seemed to work just fine - but yell if you see any issues. Thanks!

Reply:
This one did have me thinking, so I built a new image ... made sure to add a new package, and pulled in the latest feeds (to make sure the build had some differences). Did a sysupgrade (web interface), and it installed and rebooted just fine.

I admit, not sure I can explain it, but it seems to be working 😆. I could adjust the size if needed, but doesn't seem to be required. Hopefully someone smarter than me can explain ... 😄.

Reply:
If it works that's good. Maybe it's because mt7620 subtarget uses 4k sector size?
I'd be happy if there's a way to integrate loader and firmware partition, but I can't think of any good option.

Reply:
That may very well be it - I admit, not sure 😞. Thanks!

Reply:
If you are doing sysupgrade from a snapshot to snapshot that is exactly the same version, it might actually just do nothing but reboot. Watch serial log carefully just to be sure, if you see u-boot within seconds it is likely not writing to flash. Otherwise I don't see a problem with the partition size or range. As far as I understand, MTD blocks have no limitation for needing to be aligned with a sector.

Also, you might as well try getting rid of the loader partition, and putting the firmware partition back at block start, as @xabolcs  suggests. It might work since you have the correct offset finally. (I did suggest trying to revert that in forum, but it got mixed up with everything else)

Reply:
OK, that makes sense on the size. FYI, I purposely updated my feeds (LuCI), and added a new command - to make sure the code was different ... update worked fine. And it took a while to program - as well, the new command was available afterwards 😄.

Hmmm ... are you saying to drop the factory.bin, go back to sysupgrade.bin with the loader included? Want to make sure I understand. Thanks!

Reply:
Only if it works/boots

factory.bin should still exist for semantics (every device has both), but you would make them effectively the same, only difference would be no `append-metadata`.

Reply:
BTW a simpler way to test a sysupgrade success: edit the version and version.date files at the top of the build tree

Reply:
Sorry, want to make sure I do this right 😄. So remove the loader partition completely, put firmware back to 0x50000 (and adjust the length) ... right? I think boot address is still the same then. And every sysupgrade re-flashes the loader, correct?

Reply:
Right however the debate still stands as to what the standard *should* be...

In my opinion, "firmware" strictly means kernel + rootfs...
but this is not for me to decide

Reply:
Nor me, that's the one thing I know in all this ... LOL! I'm OK either way, but how to conclude? I admit, I'd like to get this (long + long standing 😆) PR in place .... device supported, and I can clean up my local branches then.

Thanks!

Reply:
And BTW, please don't take that as a complaint - it's quite the opposite ... I have learned a lot here, and I really think this answer (loader-okli) is awesome!

Reply:
> Maybe it's because mt7620 subtarget uses 4k sector size?

BTW, wanted to follow up on this. From the part datasheet (8 MB part),
![image](https://user-images.githubusercontent.com/8756384/86523350-dafddf80-be30-11ea-8076-b09fae177aed.png)

So it is a 4k sector ... but is the smallest element for reprogramming 256 bytes (page)? Either way it looks to be OK, just wanted to follow up and get back to you.

Reply:
Usually for NOR flash...
 - programming instruction write size is 256 bytes (1 page, and 256 pages per eraseblock)
 - smallest erasable sector is 4k (16 pages)
 - it can start reading from anywhere

MTD is virtualized so I'm pretty sure none of this matters anyway. It's just more conventional to have it start on a block

Reply:
Oh and it has nothing to do with target or subtarget in the build system...

linux kernel either has this info already, or it is patched directly to the kernel.
When the kernel identifies the chip, it knows what instructions are available to use on it.

for example:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/patches-4.14/302-spi-nor-add-gd25q512.patch